### ISSUE TITLE
Invalid array access in transform()

### ISSUE THREAD

**Issue by msedwar** [lmcinnes/umap#89]
When using the `fit()` then `transform()` workflow, umap shows undefined behavior such as intermittently returning a matrix of all `NaN` values, segfaulting, or even returning a correct value.

Looking into the issue, it appears as though the `transform()` function shows the undefined behavior when a training data set is used to `fit()` the model and a test set (with a cardinality greater than the training set) is passed to the `transform()` function.

The culprit appears to be the modulus `n_vertices` in the `optimize_layout()` function. When the size of `tail_embedding` is less than `n_vertices`, the accelerated code may perform an invalid array access (leading to segfaults, occassionally the correct value, etc.).

I believe the problem can be fixed by passing to `optimize_layout()` the shape of the original embedding in the `transform()` function instead of the shape of the new transformation data.

I am currently unsure if this would still be valid for the algorithm (though it still seems like a valid use case). I would appreciate any feedback on this discovery, thanks!

**Comment by lmcinnes** [comment]
Thanks that's good spotting and I believe a correct analysis of the issue.
I'll have a look to be sure and see if I can get a robust fix in place.

**Comment by jc-healy** [comment]
Great analysis of the issue.  It is a very valid use case.  I ran into this exact problem yesterday and was going to spend some time over the weekend tracking down what was causing it.  Thank you very much for saving me the time and effort.

**Comment by lmcinnes** [comment]
So I believe I just pushed what I believe is a fix for this, if either of
you can verify that the current master now works I would appreciate it.

Sorry for the delay -- I've been busy with a few other things.

**Comment by msedwar** [comment]
It appeared to still have an issue when using `fit`, then `transform`. I submitted a fix for it (the similar fix to `fit`, but for the `transform` function, and again only one line). I can verify that it works on my local.

**Comment by tbloch1** [comment]
I seem to be encountering this problem too. I'm running a 8193x6 dataset to `fit` then `transform`, which is working fine. Then I am trying to `transform` a ~43000x6 dataset and it's just returning NaNs. When I take a subset of the data (<8193, arbitrarily picked slice) it works fine again. Once I get to around 8734x6 the `transform` almost exclusively returns NaNs.

**Comment by lmcinnes** [comment]
This seems like a quirky bug :-( I did manage to fix some issues with NaN results, so I'm hoping that this turns out to be similar or related. I'll try to take a look at this on Monday.

**Comment by lmcinnes** [comment]
@tbloch1 I believe I pushed a commit for another issue that may have also fixed this one. If you could test the latest master and see if you still get the same problem I would appreciate it.

**Comment by tbloch1** [comment]
It appears to be working fine now!

**Comment by lmcinnes** [comment]
That's great news. I'm glad this worked. Hopefully this is the last of
these issues to track down.

**Comment by benaisc** [comment]
@lmcinnes Hello, we stumbled upon on the same NaNs error using transform over larger dataset than the one used to fit.
Looking at your previous fix, it seems that somehow it was rolled back ?
Maybe you ought to take a look at this :)
Thanks for your work guys, I very much appreciate this library !
